### PR TITLE
.unique() and .unique_by(..) operations

### PR DESCRIPTION
**Pull request by stepancheg**
Functions make iterators that return unique elements, storing visited
elements in a hash set.

`.unique_by(..)` uses supplied function to make keys, and `.unique()`
uses `.clone()`.

### CODE REVIEW

**Comment by stepancheg** [comment]
Updated PR with explicit test.

**Comment by bluss** [comment]
Thanks! You're taking over the task from PR #30, I guess.

I'll leave some comments inline. Style wording, etc can all be fixed by me later if you don't want to, but incorrect size hint, that won't pass :-)

File: src/adaptors.rs [hunk 1]
```diff
@@ -972,3 +974,55 @@ impl<I> Iterator for Combinations<I> where I: Iterator + Clone, I::Item: Clone{
...
+                        return Some(v);
+                    }
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
```
Comment (bluss) [inline src/adaptors.rs]: This should use the upper bound of the adapted iterator, and use a minimum of 0. Unless the hashset is empty, then use either 0 or 1 as lower bound depending on the adapted itertator.

Simply put -- our full information about how many elements remain, keeping in mind that they may all be duplicates or all unique.

File: src/lib.rs [hunk 2]
```diff
@@ -625,6 +628,45 @@ pub trait Itertools : Iterator {
...
+    /// let data = vec![10, 20, 30, 20, 40, 10, 50];
+    /// itertools::assert_equal(data.into_iter().unique(),
+    ///                         vec![10, 20, 30, 40, 50]);
+    /// ```
+    fn unique(self) -> Unique<Self> where
+        Self: Sized,
+        Self::Item: Clone + Eq + Hash,
+    {
+        self.unique_by(Clone::clone)
+    }
```
Comment (bluss) [inline src/lib.rs]: The closureless version should be before the other version

File: src/lib.rs [hunk 3]
```diff
@@ -625,6 +628,45 @@ pub trait Itertools : Iterator {
         Coalesce::new(self, eq)
     }
 
+    /// Remove non-unique elements from the iterator.
```
Comment (bluss) [inline src/lib.rs]: I think it's better to formulate it not as removing, but maybe filtering. Something about producing an iterator of unique elements (for the whole sequence).

File: src/lib.rs [hunk 4]
```diff
@@ -625,6 +628,45 @@ pub trait Itertools : Iterator {
...
+    /// Elemens are considered the same if supplied function returns
+    /// equal values for them. Those values are stored in a hash set in
+    /// the iterator.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// let data = vec!["qux", "corge", "bar", "quux", "bazaar"];
+    /// itertools::assert_equal(data.into_iter().unique_by(|s| s[..2].to_string()),
+    ///                         vec!["qux", "corge", "bar"]);
```
Comment (bluss) [inline src/lib.rs]: I can change this later myself, but I'm allergic to quux.

File: src/adaptors.rs (hunk 1 above)
Comment (stepancheg) [inline src/adaptors.rs]: Thanks, I made a mistake.

I think 0 is fine for lower bound. I think in 99% cases `size_hint` is used on fresh new iterator, so hash-set would be empty.

**Comment by stepancheg** [comment]
Changed PR according to your comments.

I didn't see #30 before you pointed to it. It's funny I made the same request independently.

**Comment by bluss** [comment]
You managed to hit the requirements very well without having seen that PR then!

Thanks for the contribution

**Comment by bluss** [comment]
A possible follow up might be: A method to extract the hash set. You might as well allow it to be used after it was constructed? Not sure.

**Comment by bluss** [comment]
Released in 0.3.16.

**Comment by Wilfred** [comment]
Marvellous! Thanks for doing this :)

**Pull request merged**